Fixes #28664 - support API middleware response callbacks

Ron-Lavi opened the pull request:
with the API responses callbacks the consumer won't need to wait until the component will get updated by the store, they can simply trigger any other action or method they want.

## Error handling

In case you want to invoke some method or action immediately when an error occurs,
we support the `handleError` param where you could pass an error handling function, e.g:
```.js
dispatch(
  get({
    key: MY_SPECIAL_KEY,
    url,
    handleError: error => error.response.status === 401 ? logoutUser() : null
    ...
  })
);
```
The error handling function will receive the error object from the API request.

## Success handling

In case you want to invoke some method or action immediately after the request succeed,
we support the `handleSuccess` param where you could pass a success handling function, e.g:
```js
dispatch(
  get({
    key: MY_SPECIAL_KEY,
    url,
    handleSuccess: response => { dispatch(addToast({ type: 'success' })) }, 
    ...
  })
);
```
The success handling function will receive the response object from the API request.

amirfefer commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -27,5 +27,9 @@ export const get = async (
       response: error,
     });
+
+    if (error.response.status === 401) {
```
The server may return with an unauthorized status in some cases but without the need for a redirection to the login page (i.e when admin permissions are required) so these cases should be filtered out, or this error handling should be part of a specific action

amirfefer reviewed (changes requested):
Thanks @laviro ! 
I think maybe this error handling should be managed by the API middleware and be defined inside an action, something like this?
```js
{
  type: API_OPERATIONS.GET,
  payload: {
    key: NOTIFICATION
    url,
    payload,
    errorHandling:  error  => {
      if (error.response.status === 401) visit('/users/login');
    } 
  }
```

amirfefer commented:
@laviro what's the status here?

amirfefer commented:
This comment is relevant for here 
https://github.com/theforeman/foreman/pull/7338/files#r391642824

Ron-Lavi commented:
thanks @amirfefer ! updated with the `handleError` param as we discussed
to create a more generic handler for such cases.

amirfefer commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -21,6 +30,7 @@ export const get = async (
       response: data,
     });
   } catch (error) {
+    handleError(error);
```
could you please add a test for this?

Ron-Lavi replied:
thanks, added

amirfefer reviewed (changes requested):
Thanks, @laviro  :+1:

ronlavi2412 committed:
Fixes #28664 - API middleware `handleError` param

the handleError callback will get called on the API request error.
the consumer won't need to wait until the component
will get updated by the store, and can invoke any method they want.

Ron-Lavi commented:
rebased and added a test

ronlavi2412 committed:
Fixes #28664 - API middleware `handleSuccess` param

Ron-Lavi commented:
added `handleSuccess` too

Ron-Lavi commented:
ready for another review

MariaAga commented:
Great addition to the api! One question: @laviro Why is the request always calling the handle function? wouldn't it be better to only call it if it was defined and give it a `null` value?

Ron-Lavi commented:
Thanks @MariaAga !
[setting those handlers initially to `noop`](https://github.com/theforeman/foreman/pull/7337/files#diff-958832b8df41edaad1c6dcf74e3ac31cR13) allow us to call them without checking

Ron-Lavi commented:
@amirfefer @sharvit could you take a look? thanks

Ron-Lavi commented:
Thanks @MariaAga,
@sharvit can you take a look?

Ron-Lavi commented:
rebased, there are several PRs that are waiting on this,
@amirfefer @jeremylenz can you maybe take a final look? Thanks

jeremylenz commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -13,14 +23,16 @@ export const get = async (
   try {
-    const { data } = await API.get(url, headers, params);
+    const response = await API.get(url, headers, params);
+    handleSuccess(response);
     dispatch({
       type: SUCCESS,
```
Seems to me like the callbacks should get called after the actions (in other words, switch the order of the `handleSuccess` and `dispatch`.)  I wouldn't want a side effect to happen before I'm even notified of what caused the side effect.

jeremylenz reviewed (commented):
Thanks @laviro !  Tested with Katello subscriptions page, works as expected.  Just one comment below

Ron-Lavi commented:
Thanks @jeremylenz ! makes sense, changed the order of the callbacks and the action dispatch

jeremylenz reviewed (approved):
Changes look good, thanks @laviro !

Ron-Lavi commented:
@amirfefer can we merge this?

amirfefer reviewed (approved):
Thanks @laviro ! Great addition  :+1:

johnpmitsch commented:
This seems to be affecting Katello tests (since they have been updated to use foremanReact directly). It shouldn't in theory, as the imports seem to be fine and webpack handles it fine, but Jest seems to have a problem with the imports in this area of code. Historically, I also had to [remove internal usage of deprecated code](https://github.com/theforeman/foreman/commit/5d07be849b7498fc7a0a311b65600b5e246645b7#diff-25ebde132b3bd4206343cbd4254b9a93) from here as it was causing a very similar issue. 

Here is the message, notice the line numbers displayed don't actually exist. I'm not sure exactly what is happening. Does anyone have any ideas?

```
FAIL webpack/scenes/ContentViews/__tests__/contentViewPage.test.js
  ● Test suite failed to run

    TypeError: Cannot read property 'APIMiddleware' of undefined



      at Object.get [as APIMiddleware] (../foreman/webpack/assets/javascripts/react_app/redux/API/index.js:28:27)
      at Object.APIMiddleware (../foreman/webpack/assets/javascripts/react_app/redux/middlewares/index.js:9:17)
      at Object.<anonymous> (../foreman/webpack/assets/javascripts/react_app/redux/index.js:10:46)
      at Object.<anonymous> (../foreman/webpack/assets/javascripts/foreman_navigation.js:4:1)
      at Object.<anonymous> (../foreman/webpack/assets/javascripts/foreman_tools.js:12:1)
      at Object.<anonymous> (../foreman/webpack/assets/javascripts/react_app/common/helpers.js:4:1)
      at Object.<anonymous> (../foreman/webpack/assets/javascripts/react_app/redux/API/APIRequest.js:3:1)
      at Object.<anonymous> (../foreman/webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js:2:1)
      at Object.<anonymous> (../foreman/webpack/assets/javascripts/react_app/redux/API/index.js:7:1)
      at Object.<anonymous> (webpack/test-utils/react-testing-lib-wrapper.js:7:1)
      at Object.<anonymous> (webpack/scenes/ContentViews/__tests__/contentViewPage.test.js:3:1)
```

This change is a workaround to the issue:
```diff
diff --git a/webpack/assets/javascripts/react_app/redux/API/APIRequest.js b/webpack/assets/javascripts/react_app/redux/API/APIRequest.js
index a97d885..f3c37cb 100644
--- a/webpack/assets/javascripts/react_app/redux/API/APIRequest.js
+++ b/webpack/assets/javascripts/react_app/redux/API/APIRequest.js
@@ -1,6 +1,7 @@
 import { API } from './';
 import { actionTypeGenerator } from './APIActionTypeGenerator';
-import { noop } from '../../common/helpers';
+//import { noop } from '../../common/helpers';
+const noop = () => {};
 
 export const get = async (
   {
```
also so does
```diff
diff --git a/webpack/assets/javascripts/react_app/common/helpers.js b/webpack/assets/javascripts/react_app/common/helpers.js
index b75fa0b..904c491 100644
--- a/webpack/assets/javascripts/react_app/common/helpers.js
+++ b/webpack/assets/javascripts/react_app/common/helpers.js
@@ -1,7 +1,8 @@
 import { snakeCase, camelCase, debounce } from 'lodash';
 import URI from 'urijs';
 import { translate as __ } from './I18n';
-import { foremanUrl } from '../../foreman_tools';
+//import { foremanUrl } from '../../foreman_tools';
+const foremanUrl = "http://localhost"
 
 /**
  * Our API returns non-ISO8601 dates
```
which leads me to believe jest is having trouble using  anything outside of `webpack/assets/javascripts/react_app`, in this case being `foreman_tools`

Unfortunately we can't stub at this level, we can only stub at Katello's entrypoint (APIMiddleware) and not further down the stack where foreman uses foreman_tools. It would be like trying to stub an internal function in an npm package rather than the external function used in the application itself.

I'm not sure the fix yet, but I'll keep looking and wanted to share my investigation so far. Please comment if you have any ideas!

Ron-Lavi commented:
Thanks @johnpmitsch for bringing that up!
I opened https://github.com/theforeman/foreman/pull/7737 to run katello JS tests on travis so we could catch such problems before the PR is merged.

as for the jest issue I moved that `foremanUrl` method into the ReactApp
and than export it also from `foreman_tools`

I wonder if we should also create an alias in tests for `foremanWebpack` than we will have access to all of the bundle.js code.

Ron-Lavi commented:
Added the fix also in https://github.com/theforeman/foreman/pull/7739,
in case #7737 doesn't fed merge

johnpmitsch commented:
@laviro  It looks like there is actually a circular dependency here:

1. This is the entry point we import `APImiddleware` from:
https://github.com/theforeman/foreman/blob/59606795ec897811de785d655a1461fd42253293/webpack/assets/javascripts/react_app/redux/API/index.js#L7
2. Which imports `get` from `APIRequest`
https://github.com/theforeman/foreman/blob/59606795ec897811de785d655a1461fd42253293/webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js#L2
3.  Which imports `noop` from `common/helpers`
https://github.com/theforeman/foreman/blob/59606795ec897811de785d655a1461fd42253293/webpack/assets/javascripts/react_app/redux/API/APIRequest.js#L3
4. Which imports `foremanUrl` from `foreman_tools`
https://github.com/theforeman/foreman/blob/59606795ec897811de785d655a1461fd42253293/webpack/assets/javascripts/react_app/common/helpers.js#L4
5. Which imports from `foreman_navigation`
https://github.com/theforeman/foreman/blob/59606795ec897811de785d655a1461fd42253293/webpack/assets/javascripts/foreman_tools.js#L12
6. Which imports from `react_app/redux`:
https://github.com/theforeman/foreman/blob/59606795ec897811de785d655a1461fd42253293/webpack/assets/javascripts/foreman_navigation.js#L4
7. Which imports `APIMiddleware` (what we were originally importing) but this time from `middlewares` instead of `API`.
https://github.com/theforeman/foreman/blob/59606795ec897811de785d655a1461fd42253293/webpack/assets/javascripts/react_app/redux/index.js#L8
8. This then calls the same `redux/API/index.js` file we originally called in step 1:
https://github.com/theforeman/foreman/blob/59606795ec897811de785d655a1461fd42253293/webpack/assets/javascripts/react_app/redux/middlewares/index.js#L1
9. Go to step 1

It seems the `redux/middlewares` export of `APIMiddleware` is there for convenience. We do use it in Katello, but can update it if it needs to be removed, I'm not sure how frequently it is used  in other plugins though. I think we will keep seeing this issue if not addressed though, perhaps it even could cause OOM issues or similar, so it's worth fixing permanently IMO.

As far as why we see this only in testing, my understanding is some build systems are better than others at fixing circular dependencies, the way webpack and jest are treating this could be different. This is about as far as my knowledge goes though, but a circular dependency does seem to be the cause of the issue.
